packages/amd-loader/src/amd-loader.ts

```diff
@@ -0,0 +1,242 @@
+
+  function onDepResolved() {
+    numUnresolvedDeps--;
+    checkDeps();
+  }
+
+  function checkDeps() {
+    if (numUnresolvedDeps === 0) {
+      if (callback !== undefined) {
+        callback.apply(null, args);
+      }
+    }
+  }
+
+  for (let i = 0; i < deps.length; i++) {
+    const depSpec = deps[i];
+
+    if (depSpec === 'exports') {
+      args[i] = mod.exports;
+
+    } else if (depSpec === 'require') {
+      args[i] = (deps: string[], callback: modCallback) =>
+          require(mod, deps, callback);
+
+    } else if (depSpec === 'meta') {
+      args[i] = {url: mod.url};
+
+    } else {
+      const depMod = getModule(resolveUrl(mod.urlBase, depSpec));
+      args[i] = depMod.exports;
+
+      if (depMod.resolved === false) {
+        numUnresolvedDeps++;
+        depMod.notify.push(onDepResolved);
+        loadIfNeeded(mod);
+      }
+    }
+  }
+
+  checkDeps();
+}
+
+let pendingDefine: ((mod: Module) => void)|undefined = undefined;
+
+/**
+ * Load a module by creating a <script> tag in the document <head>, unless we
+ * have already started (or didn't need to, as in the case of top-level
+ * scripts).
+ */
+function loadIfNeeded(mod: Module) {
+  if (mod.needsLoad === true) {
+    return;
+  }
+  mod.needsLoad = false;
+
+  const script = document.createElement('script');
+  script.src = mod.url;
+
+  script.onload = () => {
+    if (pendingDefine !== undefined) {
+      pendingDefine(mod);
+    } else {
+      // The script did not make a call to define(), otherwise the global
+      // callback would have been set. That's fine, we can resolve immediately
+      // because we don't have any dependencies, by definition.
+      define(mod, []);
+    }
+  };
+
+  script.onerror = () => {
+    throw new Error('error loading module ' + mod.url);
+  };
+
+  document.head.appendChild(script);
+}
+
+let topLevelScriptIdx = 0;
+let previousTopLevelUrl: string|undefined = undefined;
+
+/**
+ * Define a module and execute its factory function when all dependencies are
+ * resolved.
+ *
+ * Dependencies must be specified as URLs, either relative or fully qualified
+ * (e.g. "../foo.js" or "http://example.com/bar.js" but not "my-module-name").
+ */
+window.define = function(deps: string[], factory?: modCallback) {
+  // We don't yet know our own module URL. We need to discover it so that we
+  // can resolve our relative dependency specifiers. There are two ways the
+  // script executing this define() call could have been loaded:
+
+  // Case #1: We are a dependency of another module. A <script> was injected
+  // to load us, but we don't yet know the URL that was used. Because
+  // document.currentScript is not supported by IE, we communicate the URL via
+  // a global callback. When finished executing, the "onload" event will be
+  // fired by this <script>, which will be handled by the loading script,
+  // which will invoke the callback with our module object.
+  let defined = false;
+  pendingDefine = (mod) => {
```

Comment:
Talked about this offline, but I was concerned here that this was overwriting a global singleton and that doing so invites, potentially, a tricky-to-diagnose error condition if there is a possibility of two consecutive `define()` invocations before the triggering of the `pendingDefine()` inside `script.onload`.